FreeRTOS_IP.c

```diff
@@ -3889,53 +3868,157 @@ uint32_t FreeRTOS_GetIPAddress( void )
 }
 /*-----------------------------------------------------------*/
 
-#if 0
+#if ( ipconfigCOMPATIBLE_WITH_SINGLE != 0 )
 
 /*
- * The helper functions here below can not exist in a multi-interface environment.
+ * The helper functions here below assume that there is a single
+ * interface and a single end-point (ipconfigCOMPATIBLE_WITH_SINGLE)
+ */
+
+/**
+ * @brief Set the IP-address of device.
+ *
+ * @param ulIPAddress: The new IP-address.
  */
     void FreeRTOS_SetIPAddress( uint32_t ulIPAddress )
     {
         /* Sets the IP address of the NIC. */
-        *ipLOCAL_IP_ADDRESS_POINTER = ulIPAddress;
+        NetworkEndPoint_t * pxEndPoint = FreeRTOS_FirstEndPoint( NULL );
+
+        if( pxEndPoint != NULL )
+        {
+            pxEndPoint->ipv4_settings.ulIPAddress = ulIPAddress;
+        }
     }
 /*-----------------------------------------------------------*/
 
+/**
+ * @brief Get the gateway address of the subnet.
+ *
+ * @return The IP-address of the gateway, zero if a gateway is
+ *         not used/defined.
+ */
     uint32_t FreeRTOS_GetGatewayAddress( void )
     {
-        return xNetworkAddressing.ulGatewayAddress;
+        uint32_t ulIPAddress = 0U;
+        NetworkEndPoint_t * pxEndPoint = FreeRTOS_FirstEndPoint( NULL );
```

Comment:
Would returning a gateway address based on the endpoint and interface be better rather than returning the first endpoint always?

Reply:
Absolutely! But this function is declared only for backward compatibility, i.e. when `ipconfigCOMPATIBLE_WITH_SINGLE` is defined.
In that mode, there is only 1 interface, which has 1 IPv4 end-point. `ipv4_settings.ulGatewayAddress` is either a valid IP-address of a router, or it is 0.0.0.0.